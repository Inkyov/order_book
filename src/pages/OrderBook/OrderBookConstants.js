import keyMirror from 'key-mirror-nested';

export const MARKET_TICK_SIZES = {
  PI_XBTUSD: ['0.5', '1', '2.5'],
  PI_ETHUSD: ['0.05', '0.1', '0.25'],
};

export const ORDER_BOOK_TEST_IDS = keyMirror({
  TRADING_PAIRS_SELECT: null,
  TABLE_HEADERS: null,
  BIDS: null,
  ASKS: null,
});

export const COMPONENT_TEST_IDS = keyMirror({
  PRICE: null,
  TOTAL: null,
  SIZE: null,
  ORDER_ROW: null,
});
