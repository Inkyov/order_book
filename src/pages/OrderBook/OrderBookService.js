import { filter, map, isEmpty, find, reduce, includes, slice } from 'lodash/fp';

export function subscribe({ ws, stream }) {
  return ws.send(
    JSON.stringify({
      event: 'subscribe',
      feed: 'book_ui_1',
      product_ids: [stream],
    })
  );
}

export function unsubscribe({ ws, stream }) {
  return ws.send(
    JSON.stringify({
      event: 'unsubscribe',
      feed: 'book_ui_1',
      product_ids: [stream],
    })
  );
}

export const calculateTotal = (data, entry, index) => {
  const levelSize = entry[1];
  const levelSizesToSum = slice(0, index, data);
  const sumOfSizes = reduce((sum, level) => sum + level[1], 0, levelSizesToSum);

  return levelSize + sumOfSizes;
};

export const calculateDepth = (data, entry, index) => {
  const priceLevelTotal = calculateTotal(data, entry, index);
  const maxTotal = reduce((sum, level) => sum + level[1], 0, data);

  return parseFloat((priceLevelTotal / maxTotal) * 100).toFixed(2);
};

// export const calculateDepth = (total, orderBookTotal) =>
//   parseFloat((total / orderBookTotal) * 100).toFixed(2);

export const transformData = (priceLevels, delta) => {
  // an array with the already existing price levels
  const oldPriceLevels = reduce(
    (arr, priceLevel) => {
      arr.push(priceLevel[0]);
      return arr;
    },
    [],
    priceLevels
  );

  // the price levels from the delta that need to be added to the current ones
  const newPriceLevels = reduce(
    (arr, priceLevel) => {
      // if the level from the delta is not part of the current price levels it should be added
      if (!includes(priceLevel[0], oldPriceLevels)) {
        arr.push(priceLevel);
        return arr;
      }
      return arr;
    },
    [],
    delta
  );

  // overwrite the existing price level with new data
  const overwrittenPriceLevels = map((priceLevel) => {
    const newPriceLevel = find(
      (priceLevelDelta) => priceLevel[0] === priceLevelDelta[0],
      delta
    );

    // if there isn't an update for the price level return it as it is otherwise return the updated price level
    return !newPriceLevel ? priceLevel : newPriceLevel;
  }, priceLevels).concat(newPriceLevels); // add new price levels

  // remove the ones that have a size of 0 and return the rest
  const updatedPriceLevels = isEmpty(delta)
    ? priceLevels
    : filter((priceLevel) => priceLevel[1] !== 0, overwrittenPriceLevels);

  return updatedPriceLevels.reduce((arr, priceLevel, index) => {
    const total = calculateTotal(updatedPriceLevels, priceLevel, index);
    arr.push([priceLevel[0], priceLevel[1], total]);
    return arr;
  }, []);
};

export const formatPrice = (price) => parseFloat(price).toFixed(2);

export const getBids = (currentPriceLevels, delta, tickSize) =>
  transformData(currentPriceLevels, delta, tickSize).sort(
    (a, b) => b[0] - a[0]
  ); // sort in descending order

export const getAsks = (currentPriceLevels, delta, tickSize) =>
  transformData(currentPriceLevels, delta, tickSize).sort(
    (a, b) => a[0] - b[0]
  ); // sort in ascending order
