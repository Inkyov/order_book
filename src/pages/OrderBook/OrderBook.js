import React, { useState } from 'react';
import { w3cwebsocket as W3CWebSocket } from 'websocket';
import { styled, Grid, Button, Select, MenuItem } from '@material-ui/core';
import { ORDER_BOOK_TEST_IDS, MARKET_TICK_SIZES } from './OrderBookConstants';
import { subscribe, getBids, getAsks, unsubscribe } from './OrderBookService';
import AsksWidget from './components/AsksWidget';
import BidsWidget from './components/BidsWidget';

const ws = new W3CWebSocket(`wss://www.cryptofacilities.com/ws/v1`);

//eslint-disable-next-line
const Container = styled(Grid)(({ theme }) => ({
  flexDirection: 'column',
  textAlign: 'center',
  padding: 20,
  color: theme.palette.text.primary,
}));

const ToggleFeed = styled(Button)({
  backgroundColor: '#667fff',
  textTransform: 'none',
  paddingLeft: '15px',
  paddingRight: '15px',
});

const KillFeed = styled(Button)(({ theme }) => ({
  backgroundColor: theme.palette.error.dark,
  textTransform: 'none',
  paddingLeft: '15px',
  paddingRight: '15px',
}));

const OrderBook = () => {
  // eslint-disable-next-line
  const [currentStream, setCurrentStream] = useState('PI_XBTUSD');
  const [tickSize, setTickSize] = useState('0.5');
  const [feedStoped, setFeedStoped] = useState(false);
  const [bids, setBids] = useState([]);
  const [asks, setAsks] = useState([]);

  ws.onopen = () => {
    subscribe({ ws, stream: currentStream });
  };

  ws.onmessage = (result = '') => {
    const { data = '' } = result;
    const information = JSON.parse(data);

    if (information.event === 'unsubscribed' && !feedStoped) {
      setBids([]);
      setAsks([]);
      return;
    }

    setBids((prevBids) => getBids(prevBids, information.bids, tickSize));
    setAsks((prevAsks) => getAsks(prevAsks, information.asks, tickSize));
  };

  const handleTickSizeChange = ({
    event: {
      target: { value: size },
    },
  }) => {
    setTickSize(size);
  };

  const handleToggleFeed = () => {
    const newStream = currentStream === 'PI_XBTUSD' ? 'PI_ETHUSD' : 'PI_XBTUSD';
    unsubscribe({ ws, stream: currentStream });
    subscribe({ ws, stream: newStream });
    setTickSize(currentStream === 'PI_XBTUSD' ? '0.05' : '0.5');
    setCurrentStream(newStream);
  };

  const handleKillFeed = () => {
    if (feedStoped) {
      subscribe({ ws, stream: currentStream });
      setFeedStoped(false);
    } else {
      unsubscribe({ ws, stream: currentStream });
      setFeedStoped(true);
    }
  };

  return (
    <Container container spacing={3}>
      <Grid container data-testid={ORDER_BOOK_TEST_IDS.TABLE_HEADERS}>
        <Grid item xs>
          Order Book
        </Grid>
        <Grid item xs>
          Spread
        </Grid>
        <Grid item xs>
          <Select value={tickSize} onChange={handleTickSizeChange}>
            {MARKET_TICK_SIZES[currentStream].map((tickSize) => (
              <MenuItem key={tickSize} value={tickSize}>
                {tickSize}
              </MenuItem>
            ))}
          </Select>
        </Grid>
      </Grid>
      <Grid container>
        <BidsWidget
          bids={bids.slice(0, 25)}
          currentStream={currentStream}
          data-testid={ORDER_BOOK_TEST_IDS.BIDS}
        />
        <AsksWidget
          asks={asks.slice(0, 25)}
          currentStream={currentStream}
          data-testid={ORDER_BOOK_TEST_IDS.ASKS}
        />
      </Grid>

      <Grid
        container
        spacing={1}
        justifyContent="center"
        data-testid={ORDER_BOOK_TEST_IDS.TRADING_PAIRS_SELECT}
      >
        <Grid item>
          <ToggleFeed size="small" onClick={handleToggleFeed}>
            Toggle Feed
          </ToggleFeed>
        </Grid>
        <Grid item>
          <KillFeed size="small" onClick={handleKillFeed}>
            Kill Feed
          </KillFeed>
        </Grid>
      </Grid>
    </Container>
  );
};

export default OrderBook;
