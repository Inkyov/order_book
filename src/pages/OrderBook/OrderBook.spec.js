import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { cleanup, render } from '@testing-library/react';

import OrderBook from './OrderBook';
import { ORDER_BOOK_TEST_IDS } from './OrderBookConstants';

const renderComponent = () => render(<OrderBook />);

describe('OrderBook', () => {
  beforeEach(cleanup);

  it('should render', () => {
    const { container } = renderComponent();

    expect(container).toMatchSnapshot();
  });
  it('should render the select for choosing a trading pair', () => {
    const { getByTestId } = renderComponent();

    const tradingPairsSelect = getByTestId(
      ORDER_BOOK_TEST_IDS.TRADING_PAIRS_SELECT
    );
    expect(tradingPairsSelect).toBeVisible();
  });
  it('should render table headers', () => {
    const { getByTestId } = renderComponent();

    const tableHeaders = getByTestId(ORDER_BOOK_TEST_IDS.TABLE_HEADERS);
    expect(tableHeaders).toBeVisible();
  });
  it('should render bids data', () => {
    const { getByTestId } = renderComponent();

    const bids = getByTestId(ORDER_BOOK_TEST_IDS.BIDS);
    expect(bids).toBeVisible();
  });
  it('should render asks data', () => {
    const { getByTestId } = renderComponent();

    const asks = getByTestId(ORDER_BOOK_TEST_IDS.ASKS);
    expect(asks).toBeVisible();
  });
});
