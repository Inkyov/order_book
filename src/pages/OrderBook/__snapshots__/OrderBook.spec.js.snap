// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderBook should render 1`] = `
<div>
  <div
    class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-1 WithStyles(ForwardRef(Grid))-root-2 MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      class="MuiGrid-root MuiGrid-container"
      data-testid="TABLE_HEADERS"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
      >
        Order Book
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
      >
        Spread
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline"
        >
          <div
            aria-haspopup="listbox"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
            role="button"
            tabindex="0"
          >
            0.5
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            tabindex="-1"
            value="0.5"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
        currentstream="PI_XBTUSD"
        data-testid="BIDS"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-3 WithStyles(ForwardRef(Grid))-root-4 MuiGrid-item MuiGrid-grid-xs-6"
            data-testid="TOTAL"
          >
            TOTAL
          </div>
          <div
            class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-3 WithStyles(ForwardRef(Grid))-root-5 MuiGrid-item MuiGrid-grid-xs-6"
            data-testid="SIZE"
          >
            SIZE
          </div>
          <div
            class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-3 WithStyles(ForwardRef(Grid))-root-6 MuiGrid-item MuiGrid-grid-xs-6"
            data-testid="PRICE"
          >
            PRICE
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
        currentstream="PI_XBTUSD"
        data-testid="ASKS"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-7 WithStyles(ForwardRef(Grid))-root-8 MuiGrid-item MuiGrid-grid-xs-6"
            data-testid="PRICE"
          >
            PRICE
          </div>
          <div
            class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-7 WithStyles(ForwardRef(Grid))-root-9 MuiGrid-item MuiGrid-grid-xs-6"
            data-testid="SIZE"
          >
            SIZE
          </div>
          <div
            class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-7 WithStyles(ForwardRef(Grid))-root-10 MuiGrid-item MuiGrid-grid-xs-6"
            data-testid="TOTAL"
          >
            TOTAL
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-justify-content-xs-center"
      data-testid="TRADING_PAIRS_SELECT"
    >
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text WithStyles(ForwardRef(Button))-root-11 MuiButton-textSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Toggle Feed
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text WithStyles(ForwardRef(Button))-root-12 WithStyles(ForwardRef(Button))-root-13 MuiButton-textSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Kill Feed
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
